benchmark/benchmark.hpp

```diff
@@ -0,0 +1,382 @@
+ *
+ * **Customization and Extension:**
+ * - New size generators can be created by implementing the `alloc_size_interface`.
+ * - New allocators can be adapted by implementing the `allocator_interface`.
+ * - Additional benchmarking scenarios can be created by extending `benchmark_interface`.
+ */
+
+#include <benchmark/benchmark.h>
+#include <random>
+#include <umf/memory_pool.h>
+#include <umf/memory_provider.h>
+
+#include "benchmark_interfaces.hpp"
+
+struct alloc_data {
+    void *ptr;
+    size_t size;
+};
+
+#define UMF_BENCHMARK_TEMPLATE_DEFINE(BaseClass, Method, ...)                  \
+    BENCHMARK_TEMPLATE_DEFINE_F(BaseClass, Method, __VA_ARGS__)                \
+    (benchmark::State & state) {                                               \
+        for (auto _ : state) {                                                 \
+            bench(state);                                                      \
+        }                                                                      \
+    }
+
+#define UMF_BENCHMARK_REGISTER_F(BaseClass, Method)                            \
+    BENCHMARK_REGISTER_F(BaseClass, Method)                                    \
+        ->ArgNames(                                                            \
+            BENCHMARK_PRIVATE_CONCAT_NAME(BaseClass, Method)::argsName())      \
+        ->Name(BENCHMARK_PRIVATE_CONCAT_NAME(BaseClass, Method)::name())       \
+        ->MinWarmUpTime(1)
+
+class fixed_alloc_size : public alloc_size_interface {
+  public:
+    unsigned SetUp(::benchmark::State &state, unsigned argPos) override {
+        size = state.range(argPos);
+        return argPos + 1;
+    }
+    void TearDown([[maybe_unused]] ::benchmark::State &state) override {}
+    size_t nextSize() override { return size; };
+    static std::vector<std::string> argsName() { return {"size"}; }
+
+  private:
+    size_t size;
+};
+
+class uniform_alloc_size : public alloc_size_interface {
+    using distribution = std::uniform_int_distribution<int64_t>;
+
+  public:
+    unsigned SetUp(::benchmark::State &state, unsigned argPos) override {
+        auto min = state.range(argPos++);
+        auto max = state.range(argPos++);
+        auto gran = state.range(argPos++);
+        if (min % gran != 0 && max % gran != 0) {
+            state.SkipWithError("min and max must be divisible by granularity");
+            return argPos;
+        }
+
+        dist.param(distribution::param_type(min / gran, max / gran));
+        multiplier = gran;
+        return argPos;
+    }
+    void TearDown([[maybe_unused]] ::benchmark::State &state) override {}
+    size_t nextSize() override { return dist(generator) * multiplier; }
+    static std::vector<std::string> argsName() {
+        return {"min size", "max size", "granularity"};
+    }
+
+  private:
+    std::default_random_engine generator;
+    distribution dist;
+    size_t multiplier;
+};
+
+// This class benchmark speed of alloc() operations.
```

Comment:
did you mean `benchmark` -> `benchmarks` ?

Reply:
done